core/plugins/org.polarsys.capella.core.projection.scenario/src/org/polarsys/capella/core/projection/scenario/topdown/TopDownExt.java

```diff
@@ -0,0 +1,399 @@
+    Object value = null;
+    int position = 1;
+
+    if ((container != null) && (reference != null) && reference.isMany()) {
+      value = container.eGet(reference);
+      if (value instanceof List<?>) {
+        position = ((List<?>) value).indexOf(message_p) + 1;
+      }
+    }
+
+    switch (position % 10) {
+    case 1:
+      return position + Messages.Scenario_First;
+    case 2:
+      return position + Messages.Scenario_Second;
+    case 3:
+      return position + Messages.Scenario_Third;
+    default:
+      return position + Messages.Scenario_Nth;
+    }
+  }
+
+  /**
+   * Retrieve for an abstract end and a related operation, instances which are related to the bound of the operation
+   * according to the end of the message For a bound of a ce, returns related parts For a bound of a fe, returns related
+   * functions, or part according to scenario kind
+   */
+  public static Collection<AbstractInstance> getRelatedInstances(boolean isSource, EObject operation,
+      ITransfo transfo_p) {
+
+    Scenario scenario = (Scenario) transfo_p.get(TransfoEngine.TRANSFO_SOURCE);
+    // On functional scenario, we return directly related function, otherwise, allocating parts
+    boolean shouldRetrieveAllocating = !TransitionHelper.getService().isFunctionalScenario(scenario);
+
+    Collection<Component> componentBounds = new LinkedHashSet<Component>();
+    Collection<AbstractInstance> instances = new LinkedHashSet<AbstractInstance>();
+
+    if (operation instanceof ComponentExchange) {
+      // Retrieve related parts for a bound of a connection
+      // (the connected part or any representing parts of the connected type)
+      ComponentExchange exchange = (ComponentExchange) operation;
+      if (isSource || !exchange.isOriented()) {
+        instances.addAll(ComponentExchangeExt.getSourceParts(exchange));
+      }
+      if (!isSource || !exchange.isOriented()) {
+        instances.addAll(ComponentExchangeExt.getTargetParts(exchange));
+      }
+
+    } else if (operation instanceof FunctionalExchange) {
+      // Retrieve related parts for a bound of a functional exchange
+      // (any representing parts of components allocating the function)
+      FunctionalExchange exchange = (FunctionalExchange) operation;
+      AbstractFunction function = isSource ? FunctionalExchangeExt.getSourceFunction(exchange)
+          : FunctionalExchangeExt.getTargetFunction(exchange);
+
+      if (function != null) {
+        if (shouldRetrieveAllocating) {
+          componentBounds.addAll(AbstractFunctionExt.getAllocatingComponents(function));
+
+          if (function instanceof OperationalActivity) {
+            for (Role role : ((OperationalActivity) function).getAllocatingRoles()) {
+              componentBounds.addAll(role.getAllocatingEntities());
+            }
+          }
+
+        } else {
+          instances.add(function);
+        }
+      }
+    } else if (operation instanceof ExchangeItemAllocation) {
```

Comment:
To propagate SequenceMessage with ExchangeItemAllocation as operation, its here. We lookup for Components implementing/using the interface in the target Architecture